bigtable-hbase-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java

```diff
@@ -328,52 +329,80 @@ public void createTable(HTableDescriptor desc) throws IOException {
     }
+    byte[][] splitKeys;
     if (numRegions == 3) {
-      createTable(desc, new byte[][]{startKey, endKey});
-      return;
-    }
-    byte[][] splitKeys = Bytes.split(startKey, endKey, numRegions - 3);
-    if (splitKeys == null || splitKeys.length != numRegions - 1) {
-      throw new IllegalArgumentException("Unable to split key range into enough regions");
+      splitKeys = new byte[][]{startKey, endKey};
+    } else {
+      splitKeys = Bytes.split(startKey, endKey, numRegions - 3);
+      if (splitKeys == null || splitKeys.length != numRegions - 1) {
+        throw new IllegalArgumentException("Unable to split key range into enough regions");
+      }
     }
-    createTable(desc, splitKeys);
+    return splitKeys;
   }
 
   /** {@inheritDoc} */
   @Override
   public void createTable(HTableDescriptor desc, byte[][] splitKeys) throws IOException {
-    CreateTableRequest.Builder builder = CreateTableRequest.newBuilder();
+    createTable(desc.getTableName(), tableAdapter.adapt(desc, splitKeys));
+  }
+
+  protected void createTable(TableName tableName, CreateTableRequest.Builder builder) throws IOException {
     builder.setParent(bigtableInstanceName.toString());
-    builder.setTableId(desc.getTableName().getQualifierAsString());
-    builder.setTable(tableAdapter.adapt(desc));
-    if (splitKeys != null) {
-      for (byte[] splitKey : splitKeys) {
-        builder.addInitialSplits(Split.newBuilder().setKey(ByteString.copyFrom(splitKey)).build());
-      }
-    }
     try {
       bigtableTableAdminClient.createTable(builder.build());
     } catch (Throwable throwable) {
-      if (Status.fromThrowable(throwable).getCode() == Status.Code.ALREADY_EXISTS) {
-        throw new TableExistsException(desc.getTableName());
-      }
-
-      throw new IOException(
-          String.format("Failed to create table '%s'", desc.getTableName().getNameAsString()),
-          throwable);
+      throw toCreateTableThrowable(tableName, throwable);
     }
   }
 
   /** {@inheritDoc} */
   @Override
-  public void createTableAsync(HTableDescriptor desc, byte[][] splitKeys) throws IOException {
+  public void createTableAsync(final HTableDescriptor desc, byte[][] splitKeys) throws IOException {
     LOG.warn("Creating the table synchronously");
-    createTable(desc, splitKeys);
+    CreateTableRequest.Builder builder = tableAdapter.adapt(desc, splitKeys);
+    createTableAsync(builder, desc.getTableName());
+  }
+
+  protected ListenableFuture<Table> createTableAsync(CreateTableRequest.Builder builder,
+      final TableName tableName) throws IOException {
+    builder.setParent(bigtableInstanceName.toString());
+    ListenableFuture<Table> future = bigtableTableAdminClient.createTableAsync(builder.build());
+    return toCreateTableSettableFuture(tableName, future);
+  }
+
+  public static ListenableFuture<Table> toCreateTableSettableFuture(final TableName tableName,
```

Comment:
Code looks good, handles re usability well. The functionality of these methods is not easy to follow without looking at the code though. Do you think renaming them a bit differently helps readability? Some thing like handle*..